src/resty/gcp/init.lua

```diff
@@ -51,50 +51,116 @@ FindApis = function(apiClass, methods, curr)
     return methods
 end
 
+local function template_expansion(str, params)
+    return (string.gsub(str, "{(.-)}", function (key)
+        return ngx.escape_uri(params[key])
+    end))
+end
+
+local function build_request(accesstoken, apiDetail, baseUrl, params, requestBody, opt)
+    local media = opt and opt.media
+    if media and not apiDetail.supportsMediaDownload then
+        error("API does not supported media download")
+    end
+
+    local mediaUpload = apiDetail.supportsMediaUpload
+
+    local req = {
+        method = apiDetail.httpMethod,
+        headers = {
+            ["Authorization"] = "Bearer " .. accesstoken.token
+        },
+        body = requestBody,
+        ssl_verify = false
+    }
+
+    -- simple APIs
+    if apiDetail.flatPath then
+        local path = template_expansion(apiDetail.flatPath, params)
+        return path, req
+    end
+
+    local path_template
+    if mediaUpload then
+        path_template = "https://storage.googleapis.com" .. apiDetail.mediaUpload.protocols.simple.path
+    else
+        path_template = baseUrl .. apiDetail.path
+    end
+
+    local path = template_expansion(path_template, params)
+    local query = {}
+
+    if media then
+        query.alt = "media"
+    end
+
+    for k, v in pairs(params) do
+        local param = apiDetail.parameters[k]
+        if not param then
+            error("invalid parameter: " .. k)
+        end
+        local location = param.location
+        if location == "query" then
+            query[k] = v
+            -- skip paths as they are already handled
+        end
+    end
+
+    if opt and opt.media then
+        if not apiDetail.supportsMediaDownload then
+            error("API does not supported media download")
+        end
+        query.alt = "media"
+    end
```

Comment:
This appears to duplicate the logic of lines 62-64 and 93-95, so I think it can be removed:

```suggestion
```